src/Perl6/Metamodel/C3MRO.nqp

```diff
@@ -97,18 +102,20 @@ role Perl6::Metamodel::C3MRO {
             }
         }
 
-        %!mro := nqp::hash(
-            'all', nqp::hash(
-                'all', @all,
-                'all_conc', @all_conc,
-                'no_roles', @no_roles,
-            ),
-            'unhidden', nqp::hash(
-                'all', @unhidden_all,
-                'all_conc', @unhidden_all_conc,
-                'no_roles', @unhidden_no_roles,
-            ),
-        );
+        $!mro_lock.protect({
+            %!mro := nqp::hash(
```

Comment:
Does the hash need to be initialized inside the protected block?   Or could it be initialized outside in a `my %mro` and then just have the attribute binding happen inside the protected block?

Reply:
The entire point of the patch is this protection. It's not about the hash alone but about attribute read/write is not a thread-safe op.

The initialization can be split in two steps, but what would be the benefit of it? Basically, two threads updating the same MRO is a rather rare case. And it's most certainly isn't involved in a hot path. But from the perspective of readability I think it is better this way.